game.py

```diff
@@ -1,41 +1,82 @@
-from random import randint
+from operator import itemgetter
+
+from random import choice, randint
 
 from exceptions import NimException
 
 
 class Bot():
+    subclasses = []
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        Bot.subclasses.append(cls)
+
+    def valid_pile_indexes(self, state):
+        return [i for i, nb_stones in enumerate(state) if nb_stones >= 1]
+
     def move(self, state):
-        raise NotImplementedError
+        if not self.valid_pile_indexes(state):
+            raise NimException(f'No valid moves on state: {state}')
+        return self.move_strategy(state)
 
 
 class Random(Bot):
     """
-    Cory's bot :)
+    Bot that always chooses a random number of stones from a random pile
     """
-    pass
+    def move_strategy(self, state):
+        pile_index = choice(self.valid_pile_indexes(state))
+        stones = randint(1, state[pile_index])
+        return {'pile': pile_index, 'stones': stones}
 
 
 class Perfect(Bot):
     """
     Hunter's bot :)
+    Temporary strategy until implemented
     """
-    pass
+    def move_strategy(self, state):
+        pile_index = choice(self.valid_pile_indexes(state))
+        return {'pile': pile_index, 'stones': 1}
 
 
 class AI(Bot):
-    pass
+    """
+    Temporary strategy until implemented
+    """
+    def move_strategy(self, state):
+        pile_index = choice(self.valid_pile_indexes(state))
+        return {'pile': pile_index, 'stones': 1}
+
+
+class Greedy(Bot):
+    """
+    Bot that always takes as many stones as possible
+    """
+    def move_strategy(self, state):
+        pile_index, stones = max(enumerate(state), key=itemgetter(1))
+        return {'pile': pile_index, 'stones': stones}
 
 
 class Garbage(Bot):
     """
     I always take one stone from the first available pile!
     Derp! :D
     """
-    def move(self, state):
-        for i, pile in enumerate(state):
-            if pile >= 1:
-                return {'pile': i, 'stones': 1}
-        raise NimException('No stones to take')
+    def move_strategy(self, state):
+        pile_index = self.valid_pile_indexes(state)[0]
+        return {'pile': pile_index, 'stones': 1}
+
+
+class Chaos(Bot):
+    """
+    Bot that chooses a random other bot strategy to make a move
+    """
+    def move_strategy(self, state):
```

Comment:
Since this one isn't trivial, let's add a docstring to explain what this method does.

Reply:
I'm not 100% what I am supposed to do here, elaborate on what the actual code in this method does?

Reply:
Here's something that may help: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html
You don't have to follow the same format, as long as the general idea is there.

Reply:
You shouldn't use the name of the temporary variable you have in your method.
The docstring should describe the contract. What can you tell about that method to a user who is reading you code and doesn't know Python?
You need to talk about what this method is supposed to do.

For example:
```
def square(n):
    """
    Takes in a number `n` and returns its square.
    """
    s = n ** 2
    return s
```